File: python/single-responsibility/violation_user_service.py
"""
VIOLATION: This class has multiple responsibilities
It handles user data, email sending, logging, validation, analytics, and database operations
"""

import re
import datetime
from typing import Optional


class ViolationUserService:
    def __init__(self, username: str, email: str):
        self.username = username
        self.email = email
        self.connection = None
    
    # User management responsibility
    def save_user(self) -> bool:
        """Save user to database"""
        print(f"Saving user: {self.username}")
        # Database save logic
        return True
    
    def update_user(self, new_username: str, new_email: str) -> bool:
        """Update user information"""
        self.username = new_username
        self.email = new_email
        print("User updated")
        return True
    
    def delete_user(self) -> bool:
        """Delete user from database"""
        print(f"Deleting user: {self.username}")
        # Database delete logic
        return True
    
    # Email responsibility (should be separate)
    def send_welcome_email(self) -> bool:
        """Send welcome email to user"""
        print(f"Sending welcome email to: {self.email}")
        # Email sending logic
        return True
    
    def send_password_reset_email(self) -> bool:
        """Send password reset email"""
        print(f"Sending password reset email to: {self.email}")
        # Email sending logic
        return True
    
    def send_notification_email(self, subject: str, message: str) -> bool:
        """Send notification email"""
        print(f"Sending notification to {self.email}: {subject}")
        # Email sending logic
        return True
    
    # Logging responsibility (should be separate)
    def log_user_activity(self, activity: str) -> None:
        """Log user activity"""
        timestamp = datetime.datetime.now().isoformat()
        print(f"[LOG {timestamp}] User {self.username} performed: {activity}")
        # Logging logic
    
    def log_error(self, error: str, context: str) -> None:
        """Log error"""
        timestamp = datetime.datetime.now().isoformat()
        print(f"[ERROR {timestamp}] {context}: {error}")
        # Error logging logic
    
    # Validation responsibility (should be separate)
    def validate_email(self) -> bool:
        """Validate email format"""
        email_pattern = r'^[^\s@]+@[^\s@]+\.[^\s@]+$'
        return bool(re.match(email_pattern, self.email))
    
    def validate_username(self) -> bool:
        """Validate username format"""
        return 3 <= len(self.username) <= 20 and self.username.isalnum()
    
    def validate_password(self, password: str) -> bool:
        """Validate password strength"""
        return (len(password) >= 8 and 
                any(c.isupper() for c in password) and
                any(c.islower() for c in password) and
                any(c.isdigit() for c in password))
    
    # Formatting responsibility (should be separate)
    def format_user_for_display(self) -> str:
        """Format user for display"""
        return f"{self.username} ({self.email})"
    
    def format_user_for_api(self) -> dict:
        """Format user for API response"""
        return {
            "username": self.username,
            "email": self.email,
            "display_name": self.format_user_for_display()
        }
    
    # Analytics responsibility (should be separate)
    def track_user_event(self, event_name: str, metadata: Optional[dict] = None) -> None:
        """Track user event for analytics"""
        timestamp = datetime.datetime.now().timestamp()
        print(f"[ANALYTICS] {self.username}: {event_name} (timestamp: {timestamp})")
        if metadata:
            print(f"  Metadata: {metadata}")
        # Analytics tracking logic
    
    def track_user_registration(self) -> None:
        """Track user registration event"""
        self.track_user_event("user_registered")
    
    def track_user_login(self, ip_address: str) -> None:
        """Track user login event"""
        self.track_user_event("user_login", {"ip_address": ip_address})
    
    # Database responsibility (should be separate)
    def connect_to_database(self) -> bool:
        """Connect to database"""
        print("Connecting to database...")
        self.connection = "database_connection"
        return True
    
    def execute_query(self, query: str) -> list:
        """Execute database query"""
        print(f"Executing query: {query}")
        # Query execution logic
        return []
    
    def close_database_connection(self) -> None:
        """Close database connection"""
        print("Closing database connection")
        self.connection = None
    
    # Notification responsibility (should be separate)
    def send_push_notification(self, message: str) -> bool:
        """Send push notification"""
        print(f"Sending push notification to {self.username}: {message}")
        # Push notification logic
        return True
    
    def send_sms_notification(self, phone: str, message: str) -> bool:
        """Send SMS notification"""
        print(f"Sending SMS to {phone}: {message}")
        # SMS sending logic
        return True


# Example usage
if __name__ == "__main__":
    user_service = ViolationUserService("john_doe", "john@example.com")
    
    # Validate user data
    if not user_service.validate_email() or not user_service.validate_username():
        print("Invalid user data")
        exit(1)
    
    # Connect to database and save user
    user_service.connect_to_database()
    user_service.save_user()
    
    # Send welcome email
    user_service.send_welcome_email()
    
    # Log activity
    user_service.log_user_activity("User registered")
    
    # Track analytics
    user_service.track_user_registration()
    
    # Display user
    print("User display:", user_service.format_user_for_display())
    
    # Close database connection
    user_service.close_database_connection()

File: rust/single-responsibility/violation_user_service.rs
// VIOLATION: This struct has multiple responsibilities
// It handles user data, email sending, logging, validation, analytics, and database operations

use std::collections::HashMap;
use chrono::{DateTime, Utc};
use regex::Regex;

#[derive(Debug, Clone)]
pub struct ViolationUserService {
    pub username: String,
    pub email: String,
    connection: Option<String>,
}

impl ViolationUserService {
    pub fn new(username: String, email: String) -> Self {
        ViolationUserService {
            username,
            email,
            connection: None,
        }
    }
    
    // User management responsibility
    pub fn save_user(&mut self) -> Result<bool, String> {
        println!("Saving user: {}", self.username);
        // Database save logic
        Ok(true)
    }
    
    pub fn update_user(&mut self, new_username: String, new_email: String) -> Result<bool, String> {
        self.username = new_username;
        self.email = new_email;
        println!("User updated");
        Ok(true)
    }
    
    pub fn delete_user(&self) -> Result<bool, String> {
        println!("Deleting user: {}", self.username);
        // Database delete logic
        Ok(true)
    }
    
    // Email responsibility (should be separate)
    pub fn send_welcome_email(&self) -> Result<bool, String> {
        println!("Sending welcome email to: {}", self.email);
        // Email sending logic
        Ok(true)
    }
    
    pub fn send_password_reset_email(&self) -> Result<bool, String> {
        println!("Sending password reset email to: {}", self.email);
        // Email sending logic
        Ok(true)
    }
    
    pub fn send_notification_email(&self, subject: &str, message: &str) -> Result<bool, String> {
        println!("Sending notification to {}: {}", self.email, subject);
        // Email sending logic
        Ok(true)
    }
    
    // Logging responsibility (should be separate)
    pub fn log_user_activity(&self, activity: &str) {
        let timestamp = Utc::now();
        println!("[LOG {}] User {} performed: {}", timestamp.format("%Y-%m-%d %H:%M:%S"), self.username, activity);
        // Logging logic
    }
    
    pub fn log_error(&self, error: &str, context: &str) {
        let timestamp = Utc::now();
        println!("[ERROR {}] {}: {}", timestamp.format("%Y-%m-%d %H:%M:%S"), context, error);
        // Error logging logic
    }
    
    // Validation responsibility (should be separate)
    pub fn validate_email(&self) -> bool {
        let email_regex = Regex::new(r"^[^\s@]+@[^\s@]+\.[^\s@]+$").unwrap();
        email_regex.is_match(&self.email)
    }
    
    pub fn validate_username(&self) -> bool {
        self.username.len() >= 3 && self.username.len() <= 20 && 
        self.username.chars().all(|c| c.is_alphanumeric() || c == '_')
    }
    
    pub fn validate_password(&self, password: &str) -> bool {
        password.len() >= 8 &&
        password.chars().any(|c| c.is_uppercase()) &&
        password.chars().any(|c| c.is_lowercase()) &&
        password.chars().any(|c| c.is_numeric())
    }
    
    // Formatting responsibility (should be separate)
    pub fn format_user_for_display(&self) -> String {
        format!("{} ({})", self.username, self.email)
    }
    
    pub fn format_user_for_api(&self) -> HashMap<String, String> {
        let mut map = HashMap::new();
        map.insert("username".to_string(), self.username.clone());
        map.insert("email".to_string(), self.email.clone());
        map.insert("display_name".to_string(), self.format_user_for_display());
        map
    }
    
    // Analytics responsibility (should be separate)
    pub fn track_user_event(&self, event_name: &str, metadata: Option<HashMap<String, String>>) {
        let timestamp = Utc::now().timestamp();
        println!("[ANALYTICS] {}: {} (timestamp: {})", self.username, event_name, timestamp);
        if let Some(meta) = metadata {
            println!("  Metadata: {:?}", meta);
        }
        // Analytics tracking logic
    }
    
    pub fn track_user_registration(&self) {
        self.track_user_event("user_registered", None);
    }
    
    pub fn track_user_login(&self, ip_address: &str) {
        let mut metadata = HashMap::new();
        metadata.insert("ip_address".to_string(), ip_address.to_string());
        self.track_user_event("user_login", Some(metadata));
    }
    
    // Database responsibility (should be separate)
    pub fn connect_to_database(&mut self) -> Result<bool, String> {
        println!("Connecting to database...");
        self.connection = Some("database_connection".to_string());
        Ok(true)
    }
    
    pub fn execute_query(&self, query: &str) -> Result<Vec<String>, String> {
        println!("Executing query: {}", query);
        // Query execution logic
        Ok(vec![])
    }
    
    pub fn close_database_connection(&mut self) {
        println!("Closing database connection");
        self.connection = None;
    }
    
    // Notification responsibility (should be separate)
    pub fn send_push_notification(&self, message: &str) -> Result<bool, String> {
        println!("Sending push notification to {}: {}", self.username, message);
        // Push notification logic
        Ok(true)
    }
    
    pub fn send_sms_notification(&self, phone: &str, message: &str) -> Result<bool, String> {
        println!("Sending SMS to {}: {}", phone, message);
        // SMS sending logic
        Ok(true)
    }
    
    // File handling responsibility (should be separate)
    pub fn export_user_to_csv(&self) -> Result<String, String> {
        let csv_data = format!("{},{}", self.username, self.email);
        println!("Exporting user to CSV: {}", csv_data);
        // CSV export logic
        Ok(csv_data)
    }
    
    pub fn backup_user_data(&self) -> Result<bool, String> {
        println!("Backing up user data for: {}", self.username);
        // Backup logic
        Ok(true)
    }
}

// Example usage - uncomment the main function to run
/*
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut user_service = ViolationUserService::new(
        "john_doe".to_string(),
        "john@example.com".to_string(),
    );
    
    // Validate user data
    if !user_service.validate_email() || !user_service.validate_username() {
        println!("Invalid user data");
        return Ok(());
    }
    
    // Connect to database and save user
    user_service.connect_to_database()?;
    user_service.save_user()?;
    
    // Send welcome email
    user_service.send_welcome_email()?;
    
    // Log activity
    user_service.log_user_activity("User registered");
    
    // Track analytics
    user_service.track_user_registration();
    
    // Display user
    println!("User display: {}", user_service.format_user_for_display());
    
    // Export to CSV
    let csv_data = user_service.export_user_to_csv()?;
    
    // Close database connection
    user_service.close_database_connection();
    
    Ok(())
}
*/